File: gtci/04-merge-intervals/python3/3-1-find-conflicting-appointments.py
import heapq


class Interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end

    def __str__(self):
        return f"[{self.start}, {self.end}]"

    def __repr__(self):
        return f"Interval({self.start}, {self.end})"


class Solution:
    def getConflictingAppointments(self, intervals):
        if len(intervals) < 2:
            return []

        heap = []
        conflicts = []

        intervals.sort(key=lambda x: x.start)

        for interval in intervals:
            # pop all intervals that end before the current one starts
            while heap and heap[0][0] <= interval.start:
                heapq.heappop(heap)

            # all remaining intervals in the heap conflict with the current interval
            for conflict in heap:
                conflicts.append(
                    [
                        [conflict[1].start, conflict[1].end],
                        [interval.start, interval.end],
                    ]
                )

            # push the current interval onto the heap
            heapq.heappush(heap, (interval.end, interval))

        return conflicts


if __name__ == "__main__":
    sol = Solution()

    assert sol.getConflictingAppointments(
        [Interval(4, 5), Interval(2, 3), Interval(3, 6), Interval(5, 7), Interval(7, 8)]
    ) == [[[3, 6], [4, 5]], [[3, 6], [5, 7]]]

    print("All test cases passed.")

File: gtci/04-merge-intervals/cpp/3-1-find-conflicting-appointments.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

#include "interval.h"

using namespace std;

class Solution {
   public:
    vector<pair<Interval, Interval>> getConflictingAppointments(
        vector<Interval>& intervals) {
        vector<pair<Interval, Interval>> conflicts;

        if (intervals.size() < 2) return conflicts;

        std::sort(intervals.begin(), intervals.end(),
                  [](Interval a, Interval b) { return a.start < b.start; });

        // interval comparator function for min-heap
        auto intervalCmp = [](const Interval& a, const Interval& b) {
            return a.end > b.end;  // smaller end comes first
        };

        std::priority_queue<Interval, std::vector<Interval>,
                            decltype(intervalCmp)>
            minHeap(intervalCmp);

        for (Interval& curr : intervals) {
            // pop all intervals that end before the current one starts
            while (!minHeap.empty() && minHeap.top().end <= curr.start)
                minHeap.pop();

            // all remaining intervals overlap
            auto temp = minHeap;
            while (!temp.empty()) {
                conflicts.push_back({temp.top(), curr});
                temp.pop();
            }

            // push the current interval into the heap
            minHeap.push(curr);
        }

        return conflicts;
    }
};

int main() {
    Solution sol;

    vector<Interval> intervals1 = {Interval(4, 5), Interval(2, 3),
                                   Interval(3, 6), Interval(5, 7),
                                   Interval(7, 8)};
    vector<pair<Interval, Interval>> expected = {
        {Interval(3, 6), Interval(4, 5)}, {Interval(3, 6), Interval(5, 7)}};
    auto result = sol.getConflictingAppointments(intervals1);
    assert(result.size() == expected.size());
    for (size_t i = 0; i < result.size(); i++) {
        assert(result[i].first == expected[i].first);
        assert(result[i].second == expected[i].second);
    }

    cout << "All test cases passed." << endl;

    return 0;
}
